sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,382 @@
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Save hits to HDF5: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed save events to HDF5.
+ * 
+ * @param[in] out_events
+ * @param[in] batches
+ */
+void timedSaveEventsToHDF5(const std::string &out_events, std::vector<TPX3> &batches) {
+  auto start = std::chrono::high_resolution_clock::now();
+  // move all events into a single vector
+  std::vector<Neutron> events;
+  for (const auto &tpx3 : batches) {
+    auto tpx3_events = tpx3.neutrons;
+    events.insert(events.end(), tpx3_events.begin(), tpx3_events.end());
+  }
+  // save events to HDF5 file
+  saveNeutronToHDF5(out_events, events);
+  auto end = std::chrono::high_resolution_clock::now();
+  auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - start).count();
+  spdlog::info("Save events to HDF5: {} s", elapsed / 1e6);
+}
+
+/**
+ * @brief Timed create TOF images.
+ *
+ * This function creates time-of-flight (TOF) images based on the provided batches of TPX3 data. The TOF images are 2D histograms that represent the distribution of neutron events in space for different TOF bins. The function takes into account the super resolution, TOF bin edges, and the mode of operation (hit or neutron) to generate the TOF images.
+ *
+ * @param[in] batches The vector of TPX3 batches containing the neutron events.
+ * @param[in] super_resolution The super resolution factor used to calculate the dimensions of each 2D histogram.
+ * @param[in] tof_bin_edges The vector of TOF bin edges used to determine the TOF bin for each neutron event.
+ * @param[in] mode The mode of operation, either "hit" or "neutron", which determines the type of events to process.
+ * @return std::vector<std::vector<std::vector<unsigned int>>> The vector of TOF images, where each TOF image is a 2D histogram representing the distribution of neutron events in space for a specific TOF bin.
+ */
+std::vector<std::vector<std::vector<unsigned int>>> timedCreateTOFImages(
+    const std::vector<TPX3>& batches, 
+    double super_resolution, 
+    const std::vector<double>& tof_bin_edges,
+    const std::string& mode) {
+    
+    auto start = std::chrono::high_resolution_clock::now();
+
+    // Initialize the TOF images container
+    std::vector<std::vector<std::vector<unsigned int>>> tof_images(tof_bin_edges.size() - 1);
+
+    // Sanity checks
+    if (tof_bin_edges.size() < 2) {
+        spdlog::error("Invalid TOF bin edges: at least 2 edges are required");
+        return {};
+    }
+    if (batches.empty()) {
+        spdlog::error("No batches to process");
+        return tof_images;
+    }
+    
+    // Calculate the dimensions of each 2D histogram based on super_resolution
+    // one chip: 0-255 pixel pos
+    // gap: 5
+    // total: 0-255 + 5 + 0-255 -> 517 (TPX3@VENUS only)
+    int dim_x = static_cast<int>(517 * super_resolution);
```

Comment:
Magic numbers are bad! Make 517 a `constexpr int` in an anonymous namespace at the top of the file.

Reply:
I am still trying to find a good centralized location for these kind of numbers, unfortunately TPX3 related processing has loads of this kind of parameters and they are all over the place.

Reply:
#58